### DISCUSSION TITLE
你真的了解 load 方法么？ · /load

### DISCUSSION THREAD

**Discussion by draveness** [draveness/blog-comments#293] [Announcements]
https://draveness.me/load

**Comment by xuzhongping** [comment]
前辈你好，关于下面这句话我有点不太理解:
但是由于 load 方法的运行时间过早，所以这里可能不是一个理想的环境，因为某些类可能需要在在其它类之前加载，但是这是我们无法保证的。不过在这个时间点，所有的 framework 都已经加载到了运行时中，所以调用 framework 中的方法都是安全的。

上面所说所有的 framework 都已经加载到了运行时中，所以调用 framework 中的方法都是安全的。那么framework以外的类在此时调用难道就不安全了么？我在源码中看到所有的类是在_read_images函数中进行注册到一个表中的，而且会在随后依次调用realizeClass函数进行类的实现(这个在你的另一篇关于方法的结构中也有说到)，然后才会调用load_images函数进行callLoadMethod的流程，此时除了运行时动态添加的类，其他所有类也都已经实现好，所以不太明白你所说的framework里的类和方法才是安全的这句话，希望能得到解惑！

**Comment by alwaystogo** [comment]
因为framework以及动态库是先加载的，所有库加载完成之后，才会执行运行时的load方法。 这是我个人的理解。
> @JungHsu
> 前辈你好，关于下面这句话我有点不太理解:
> 但是由于 load 方法的运行时间过早，所以这里可能不是一个理想的环境，因为某些类可能需要在在其它类之前加载，但是这是我们无法保证的。不过在这个时间点，所有的 framework 都已经加载到了运行时中，所以调用 framework 中的方法都是安全的。
> 
> 上面所说所有的 framework 都已经加载到了运行时中，所以调用 framework 中的方法都是安全的。那么framework以外的类在此时调用难道就不安全了么？我在源码中看到所有的类是在_read_images函数中进行注册到一个表中的，而且会在随后依次调用realizeClass函数进行类的实现(这个在你的另一篇关于方法的结构中也有说到)，然后才会调用load_images函数进行callLoadMethod的流程，此时除了运行时动态添加的类，其他所有类也都已经实现好，所以不太明白你所说的framework里的类和方法才是安全的这句话，希望能得到解惑！

**Comment by jsonsnow** [comment]
> @JungHsu
> 前辈你好，关于下面这句话我有点不太理解:
> 但是由于 load 方法的运行时间过早，所以这里可能不是一个理想的环境，因为某些类可能需要在在其它类之前加载，但是这是我们无法保证的。不过在这个时间点，所有的 framework 都已经加载到了运行时中，所以调用 framework 中的方法都是安全的。
> 
> 上面所说所有的 framework 都已经加载到了运行时中，所以调用 framework 中的方法都是安全的。那么framework以外的类在此时调用难道就不安全了么？我在源码中看到所有的类是在_read_images函数中进行注册到一个表中的，而且会在随后依次调用realizeClass函数进行类的实现(这个在你的另一篇关于方法的结构中也有说到)，然后才会调用load_images函数进行callLoadMethod的流程，此时除了运行时动态添加的类，其他所有类也都已经实现好，所以不太明白你所说的framework里的类和方法才是安全的这句话，希望能得到解惑
